### DISCUSSION TITLE
Does this support pagination? Keyset pagination?

### DISCUSSION THREAD

**Discussion by andrewbaxter** [oxigraph/oxigraph#685] [Q&A] [answered]
AFAICT looking through the rust docs it returns the whole query set.  I imagine you could implement something similar yourself, but you'd need to use a new query to do so (possibly multiple queries depending on where you left off).

**Comment by Tpt** [comment, marked as answer]
Hey! Thank you for having a look at Oxigraph.
* SPARQL [has `OFFSET` and `LIMIT` modifiers](https://www.w3.org/TR/sparql11-query/#modOffset) just like SQL. This allows basic pagination.
* The iterator returned by `Store.query` is lazy i.e. it will only load results when needed. So, if the query evaluator does not have to materialize intermediate states during evaluation there should not be problems with query results larger than RAM.
* Keyset pagination is sadly not optimized yet. Feel free to open a feature request about it.

**Reply by andrewbaxter** [reply]
Ah thanks, it makes sense that it's part of the QL... I should have thought of that!  So I guess that means the ordering of the results of a query is deterministic (modulo intervening modifications).

**Reply by Tpt** [reply]
> So I guess that means the ordering of the results of a query is deterministic (modulo intervening modifications).

Yes, with the current implementation it should. But it might change in the future.
